benchmarks/distributed/rpc/parameter_server/trainers/DdpSparseRpcTrainer.py

```diff
@@ -0,0 +1,149 @@
+        "AverageParameterServer": AverageParameterServer
+    }
+
+    class HookState:
+
+        def __init__(self, cref, process_group):
+            self.cref = cref
+            self.process_group = process_group
+            self.process_group_size = process_group.size()
+            self.param_loc = 0
+            self.batch_number = -1
+
+        def get_key(self):
+            return f"{self.batch_number},{self.param_loc}"
+
+        def next_batch_state(self):
+            self.param_loc = 0
+            self.batch_number += 1
+
+    def __init__(self, rank, trainer_count, ps_rref, ps_name, backend, use_cuda_rpc, epochs):
+        super().__init__(rank)
+        self.rank = rank
+        self.trainer_count = trainer_count
+        self.ps_rref = ps_rref
+        self.ps = self.PS_MAP[ps_name]
+        self.backend = backend
+        self.use_cuda_rpc = use_cuda_rpc
+        self.epochs = epochs
+
+    @staticmethod
+    def get_tensor_fut(bucket):
+        fut = torch.futures.Future()
+        fut.set_result([bucket.get_tensor()])
+        return fut
+
+    @staticmethod
+    def process_bucket(state, bucket):
+        cref = state.cref
+        tensor = bucket.get_tensor()
+        tensors_count = len(cref.bucket_to_parameters(bucket))
+        sparse = tensor.is_sparse
+        if not cref.use_cuda_rpc:
+            tensor = tensor.cpu()
+        if sparse:
+            tensor = cref.sparse_tensor_to_rpc_format(tensor)
+        ps = cref.ps
+        ps_args = [
+            cref.ps_rref,
+            state.batch_number,
+            state.param_loc,
+            tensor
+        ]
+        fut = cref.send_async_request(
+            state.get_key(),
+            cref.ps_rref,
+            ps.average_gradient,
+            *ps_args
+        )
+        state.param_loc += tensors_count
+
+        def callback(fut):
+            tensor = fut.wait()
+            if type(tensor) is list:
+                tensor = cref.sparse_rpc_format_to_tensor(tensor)
+            if not cref.use_cuda_rpc:
+                tensor = tensor.cuda(cref.rank)
+            return [tensor]
+
+        return fut.then(callback)
+
+    @staticmethod
+    def hook(state, bucket):
+        cref = state.cref
+        tensor = bucket.get_tensor()
+        tensors_count = len(cref.bucket_to_parameters(bucket))
+        if tensor.is_sparse:
+            return cref.process_bucket(state, bucket)
+        else:
+            tensor = [tensor / state.process_group_size]
+            key = state.get_key()
+            cref.record_hook_fut_start(key, cref.backend)
+            fut = state.process_group.allreduce(tensor).get_future()
+            state.param_loc += tensors_count
+
+            def callback(fut):
+                cref.record_hook_fut_end(key)
+                return fut.wait()
+
+            return fut.then(callback)
+
+    def get_hook(self):
+        return DdpSparseRpcTrainer.hook
+
+    def train(self, model, data):
+        torch.manual_seed(0)
```

Comment:
If we are setting a seed, then maybe this should be done in launcher? Also don't we also need to seed cuda?

Reply:
add seed to launcher. also, added backend reproducibility.
```python
    torch.manual_seed(args.torch_seed)
    torch.cuda.manual_seed_all(args.cuda_seed)
    torch.backends.cudnn.benchmark = True
    torch.backends.cudnn.deterministic = True
```